flink-dstl/flink-dstl-dfs/src/main/java/org/apache/flink/changelog/fs/UploadThrottle.java

```diff
@@ -0,0 +1,91 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.changelog.fs;
+
+import org.apache.flink.runtime.io.AvailabilityProvider;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.concurrent.GuardedBy;
+import javax.annotation.concurrent.ThreadSafe;
+
+import java.util.concurrent.CompletableFuture;
+
+/** Helper class to throttle upload requests when the in-flight data size limit is exceeded. */
+@ThreadSafe
+class UploadThrottle implements AvailabilityProvider {
+    private static final Logger LOG = LoggerFactory.getLogger(UploadThrottle.class);
+
+    private final Object lock = new Object();
+    private final long maxBytesInFlight;
+
+    @GuardedBy("lock")
+    private long inFlightBytesCounter = 0;
+
+    @GuardedBy("lock")
+    private CompletableFuture<?> availabilityFuture = AvailabilityProvider.AVAILABLE;
+
+    UploadThrottle(long maxBytesInFlight) {
+        this.maxBytesInFlight = maxBytesInFlight;
+    }
+
+    /**
+     * Seize <b>bytes</b> capacity, waiting if needed. Called by the Task thread.
+     *
+     * @throws InterruptedException
+     */
+    public void seizeCapacity(long bytes) throws InterruptedException {
+        synchronized (lock) {
+            while (!hasCapacity()) {
+                LOG.info("In flight data size threshold exceeded: {}", maxBytesInFlight);
+                lock.wait();
+            }
```

Comment:
If we want to go with this backpressured metric this code should be probably wrapped with
```
backPressuredTimeMsPerSecond.markStart();
(...)
backPressuredTimeMsPerSecond.markEnd();
```

Reply:
As discussed offline, I'll remove the use of `backPressuredTimePerSecond` in this PR, so any waiting on changelog will be reported as busy (which is consistent with other state backends currently).

The mertric then should be addressed in FLINK-24402.

Reply:
here `waiting` means blocking-waiting.
- blocking waiting is reported as busy
- non-blocking waiting is reported ?? (follow-up)